src/library/scala/collection/immutable/LazyList.scala

```diff
@@ -1370,7 +1377,7 @@ object LazyList extends SeqFactory[LazyList] {
         case a => init += a.asInstanceOf[A]
       }
       val tail = in.readObject().asInstanceOf[LazyList[A]]
-      coll = init ++: tail
+      coll = tail.withNullLazyState(tail.prependedAll(init))
```

Comment:
why don't we use a `Builder[LazyList]` for `init` instead of an `ArrayBuffer`, and just call `lazyAppendedAll`?

Reply:
in fact, because `tail` is also a `LazyList`, I believe we can use `LazyListBuilder` and just call `addAll` with `tail` as an argument, and we don't even rebuild the `LazyList` ever

Reply:
side note: there was a discussion at some point somewhere suggesting that no `Builder`s should be lazy, and that `LazyList.newBuilder` should return an eager `Builder` (probably for `List`) that calls `mapResult` to create a `LazyList`. On the off-chance that ever happens, we should explicitly create a `new LazyBuilder` rather than calling `LazyList.newBuilder` if we go with the change I suggested

Reply:
to put it differently: evaluating `tail` there does not represent a fundamental flaw in the design of `LazyList`, it represents a careless programming error likely due to copy-pasting

Reply:
Looking at `LazyBuilder`, I actually think we'd run into the same issue. If we change the code to

```
      val tail = in.readObject().asInstanceOf[LazyList[A]]
      coll = lazyBuilder.addAll(tail).result()
```

we have

```
    override def addAll(xs: IterableOnce[A]): this.type = {
      if (xs.knownSize != 0) {
        ...

  override def knownSize: Int = if (knownIsEmpty) 0 else -1
...
  private[this] def knownIsEmpty: Boolean = stateEvaluated && isEmpty
...
  override def isEmpty: Boolean = state eq State.Empty
```

So calling `addAll` could evaluate the `state` lazy val, which calls the `lazyState` function.

Reply:
`knownIsEmpty` only evaluates `isEmpty` if `stateEvaluated == true`

Reply:
Yeah, a forged serialization stream can have that field set to `true`

Reply:
point. I guess we fall back to `lazyAppendedAll` then

Reply:
@lrytz I whipped up my own fix that uses `stateFromIteratorConcatSuffix`. I was going to push an alternate PR to consider, but I haven't figured out how to actually test that the fix works. Been having issues deserializing the lambda.

If you'd like to use my code, the only changes are to the `SerializationProxy` (below)

```scala
  final class SerializationProxy[A](@transient protected var coll: LazyList[A]) extends Serializable {

    private[this] def writeObject(out: ObjectOutputStream): Unit = {
      out.defaultWriteObject()
      var these = coll
      while(these.knownNonEmpty) {
        out.writeObject(these.head)
        these = these.tail
      }
      out.writeObject(SerializeEnd)
      out.writeObject(these)
    }

    private[this] def readObject(in: ObjectInputStream): Unit = {
      in.defaultReadObject()
      val init = new ListBuffer[A]
      var initRead = false
      while (!initRead) in.readObject match {
        case SerializeEnd => initRead = true
        case a => init += a.asInstanceOf[A]
      }
      val tail = in.readObject().asInstanceOf[LazyList[A]]
      coll = newLL(stateFromIteratorConcatSuffix(init.toList.iterator)(tail.state))
    }

    private[this] def readResolve(): Any = coll
  }
```

and to import `ListBuffer` instead of `ArrayBuffer`.

Reply:
Thank you, this looks good, nicer than my workaround 👍

Reply:
Also added a test.